### DISCUSSION TITLE
Get head_comment for each document

### DISCUSSION THREAD

**Discussion by krokofant** [mikefarah/yq#2185] [Q&A] [answered]
Consider these documents in a single file
```yaml
# test1
foo: bar
---
# test2
foo: bar2
---
# test3
foo: bar3
```

I would consider all the _test1, test2, test3_ to be `head_comment`s, so:
```powershell
yq ea 'head_comment' ./input.yaml
```
 which produces the comment for the first document:
 ```
test1


```

Then to dig deeper I found this in the docs, to find where all comments are. Notice that in document 2 and 3 the head_comment is attached to the `.foo` prop :
`yq '[... | {"p": path | join("."), "isKey": is_key, "hc": headComment, "lc": lineComment, "fc": footComment}]' ./input.yaml`

```yaml
- p: ""
  isKey: false
  hc: test1
  lc: ""
  fc: ""
- p: foo
  isKey: true
  hc: ""
  lc: ""
  fc: ""
- p: foo
  isKey: false
  hc: ""
  lc: ""
  fc: ""
- p: ""
  isKey: false
  hc: ""
  lc: ""
  fc: ""
- p: foo
  isKey: true
  hc: test2
  lc: ""
  fc: ""
- p: foo
  isKey: false
  hc: ""
  lc: ""
  fc: ""
- p: ""
  isKey: false
  hc: ""
  lc: ""
  fc: ""
- p: foo
  isKey: true
  hc: test3
  lc: ""
  fc: ""
- p: foo
  isKey: false
  hc: ""
  lc: ""
  fc: ""
```

So the document 2 and 3 head comments can be extracted like `yq ea '.foo | key | head_comment' ./input.yaml`:
```

---
test2
---
test3
```

This seems like a faulty behavior to me. Can the head_comment be fetched from all documents without having to know about the document keys? 😕

**Comment by mikefarah** [comment, marked as answer]
Yeah it's definitely faulty, and a known issue.  Comment handling is something I've been trying to beat the underlying yaml parser with a stick with - with the way the parser attaches comments to various nodes it's really hard to provide a consistent intuitive behavior for all the various scenarios :(  

TLDR I had to do some head pre-processing to fix other header comment parsing issues - but this causes issues in other places, like this. You can turn off the pre-processing and get document 1 to be consistent with document 2 and 3:

```
yq --header-preprocess=f '.foo | key | head_comment'
```

Just a word of warning that without the header-preprocessing you can get some issues with block comments at the start of yaml files (especially with newlines; and document separation tags at the start/end of the comment block.
